spring-cloud-aws-sns/src/main/java/io/awspring/cloud/sns/core/TopicMessageChannel.java

```diff
@@ -0,0 +1,172 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.awspring.cloud.sns.core;
+
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import com.fasterxml.jackson.core.io.JsonStringEncoder;
+import software.amazon.awssdk.core.SdkBytes;
+import software.amazon.awssdk.services.sns.SnsClient;
+import software.amazon.awssdk.services.sns.model.MessageAttributeValue;
+import software.amazon.awssdk.services.sns.model.PublishRequest;
+
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.support.AbstractMessageChannel;
+import org.springframework.util.Assert;
+import org.springframework.util.MimeType;
+import org.springframework.util.NumberUtils;
+
+import static io.awspring.cloud.sns.core.MessageHeaderCodes.MESSAGE_DEDUPLICATION_ID_HEADER;
+import static io.awspring.cloud.sns.core.MessageHeaderCodes.MESSAGE_GROUP_ID_HEADER;
+import static io.awspring.cloud.sns.core.MessageHeaderCodes.NOTIFICATION_SUBJECT_HEADER;
+
+/**
+ * @author Agim Emruli
+ * @author Alain Sahli
+ * @author Gyozo Papp
+ * @since 1.0
+ */
+public class TopicMessageChannel extends AbstractMessageChannel {
+
+	private final JsonStringEncoder jsonStringEncoder = JsonStringEncoder.getInstance();
+
+	private final SnsClient snsClient;
+
+	private final String topicArn;
+
+	public TopicMessageChannel(SnsClient snsClient, String topicArn) {
+		this.snsClient = snsClient;
+		this.topicArn = topicArn;
+	}
+
+	private static String findNotificationSubject(Message<?> message) {
+		return message.getHeaders().containsKey(NOTIFICATION_SUBJECT_HEADER)
+				? message.getHeaders().get(NOTIFICATION_SUBJECT_HEADER).toString() : null;
+	}
+
+	@Override
+	protected boolean sendInternal(Message<?> message, long timeout) {
+		PublishRequest.Builder publishRequestBuilder = PublishRequest.builder();
+		publishRequestBuilder.topicArn(this.topicArn).message(message.getPayload().toString())
+				.subject(findNotificationSubject(message));
+		Map<String, MessageAttributeValue> messageAttributes = getMessageAttributes(message);
+		if (!messageAttributes.isEmpty()) {
+			publishRequestBuilder.messageAttributes(messageAttributes);
+		}
+		if (message.getHeaders().containsKey(MESSAGE_GROUP_ID_HEADER)) {
+			publishRequestBuilder.messageGroupId(message.getHeaders().get(MESSAGE_GROUP_ID_HEADER, String.class));
+		}
+		if (message.getHeaders().containsKey(MESSAGE_DEDUPLICATION_ID_HEADER)) {
+			publishRequestBuilder
+					.messageDeduplicationId(message.getHeaders().get(MESSAGE_DEDUPLICATION_ID_HEADER, String.class));
+		}
```

Comment:
I think this is simpler and less bug-prone:
```suggestion
		Optional.ofNullable(message.getHeaders().get(MESSAGE_DEDUPLICATION_ID_HEADER, String.class))
			.ifPresent(publishRequestBuilder::messageDeduplicationId);
```